src/PostgREST/AppState.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module PostgREST.AppState
+  ( AppState
+  , init
+  , initWithPool
+  , pool
+  , releasePool
+  , pgVersion
+  , isWorkerOn
+  , time
+  , config
+  , dbStructure
+  , putDbStructure
+  , putIsWorkerOn
+  , putPgVersion
+  , putConfig
+  , killMainThread
+  ) where
+
+import qualified Hasql.Pool as P
+
+import Control.AutoUpdate (defaultUpdateSettings, mkAutoUpdate,
+                           updateAction)
+import Data.IORef         (IORef, atomicWriteIORef, newIORef,
+                           readIORef)
+import Data.Time.Clock    (UTCTime, getCurrentTime)
+
+import PostgREST.Config                (AppConfig (..))
+import PostgREST.DbStructure           (DbStructure)
+import PostgREST.DbStructure.PgVersion (PgVersion (..))
+
+import Protolude      hiding (toS)
+import Protolude.Conv (toS)
+
+
+data AppState = AppState
+  { statePool         :: P.Pool
+  -- ^ Connection pool, either a 'Connection' or a 'ConnectionError'
+  , statePgVersion    :: MVar PgVersion
+  -- ^ Used to sync the listener(NOTIFY reload) with the connectionWorker. No
+  -- connection for the listener at first. Only used if dbChannelEnabled=true.
+  , stateDbStructure  :: IORef (Maybe DbStructure)
+  -- ^ No schema cache at the start. Will be filled in by the connectionWorker
+  , stateIsWorkerOn   :: IORef Bool
+  -- ^ Helper ref to make sure just one connectionWorker can run at a time
+  , stateConf         :: IORef AppConfig
+  -- ^ Config that can change at runtime
+  , stateGetTime      :: IO UTCTime
+  , stateMainThreadId :: ThreadId
+  }
+
+init :: AppConfig -> IO AppState
+init conf = do
+  newPool <- initPool conf
+  initWithPool newPool conf
+
+initWithPool :: P.Pool -> AppConfig -> IO AppState
+initWithPool newPool conf =
+  AppState newPool
+    <$> newEmptyMVar
+    <*> newIORef Nothing
+    <*> newIORef False
+    <*> newIORef conf
+    <*> getTime
+    <*> myThreadId
+  where
+    getTime = mkAutoUpdate defaultUpdateSettings { updateAction = getCurrentTime }
+
+
+initPool :: AppConfig -> IO P.Pool
+initPool AppConfig{..} =
+  P.acquire
+    ( configDbPoolSize
+    , fromRational . toRational $ configDbPoolTimeout
+    , toS configDbUri
+    )
+
+pool :: AppState -> P.Pool
+pool = statePool
+
+releasePool :: AppState -> IO ()
+releasePool AppState{..} = P.release statePool >> throwTo stateMainThreadId UserInterrupt
+
+pgVersion :: AppState -> IO PgVersion
+pgVersion = takeMVar . statePgVersion
+
+putPgVersion :: AppState -> PgVersion -> IO ()
+putPgVersion appState pgVer = void $ tryPutMVar (statePgVersion appState) pgVer
+
+dbStructure :: AppState -> IO (Maybe DbStructure)
+dbStructure = readIORef . stateDbStructure
```

Comment:
Commenting on a random one of those here:

So I understand we have some getters, because we need to read the `IORef`s in some cases - that's why it doesn't really help to use `RecordWildcards` or whatever at the callsite. I guess you added those other getters (`pool` looks quite simple) for consistency.

How about naming them consistently something like `getDbStructure`, `getPool` or whatever?

Reply:
I renamed all getters that need IO to `get...`. I left `pool` as is, as it does not run in IO at directly returns the pool.

Reply:
At some point `AppState` might become a monad with a few small fixes - but I think for this PR it's fine and usable as is.